File: fotoapparat/src/main/java/io/fotoapparat/parameter/selector/FocusModeSelectors.java
package io.fotoapparat.parameter.selector;

import java.util.Collection;

import io.fotoapparat.parameter.FocusMode;

/**
 * Selector functions for {@link FocusMode}.
 */
public class FocusModeSelectors {

    /**
     * @return {@link SelectorFunction} which provides a non-adjustable focus mode if available.
     * Otherwise provides {@code null}.
     */
    public static SelectorFunction<Collection<FocusMode>, FocusMode> fixed() {
        return focusMode(FocusMode.FIXED);
    }

    /**
     * @return {@link SelectorFunction} which provides a focus mode targeting infinity if available.
     * Otherwise provides {@code null}.
     */
    public static SelectorFunction<Collection<FocusMode>, FocusMode> infinity() {
        return focusMode(FocusMode.INFINITY);
    }

    /**
     * @return {@link SelectorFunction} which provides a macro focus mode if available.
     * Otherwise provides {@code null}.
     */
    public static SelectorFunction<Collection<FocusMode>, FocusMode> macro() {
        return focusMode(FocusMode.MACRO);
    }

    /**
     * @return {@link SelectorFunction} which provides an auto focus mode if available.
     * Otherwise provides {@code null}.
     */
    public static SelectorFunction<Collection<FocusMode>, FocusMode> autoFocus() {
        return focusMode(FocusMode.AUTO);
    }

    /**
     * @return {@link SelectorFunction} which provides a focus mode which constantly tries to stay
     * in focus if available.
     * Otherwise provides {@code null}.
     */
    public static SelectorFunction<Collection<FocusMode>, FocusMode> continuousFocus() {
        return focusMode(FocusMode.CONTINUOUS_FOCUS);
    }

    /**
     * @return {@link SelectorFunction} which provides a focus mode which will produce images with
     * an extended depth of field if available.
     * Otherwise provides {@code null}.
     */
    public static SelectorFunction<Collection<FocusMode>, FocusMode> edof() {
        return focusMode(FocusMode.EDOF);
    }

    /**
     * @return function which selects given {@link FocusMode} from the list if it is available. If
     * it is not available - provides {@code null}.
     */
    private static SelectorFunction<Collection<FocusMode>, FocusMode> focusMode(final FocusMode focusMode) {
        return Selectors.single(focusMode);
    }

}

File: fotoapparat/src/main/java/io/fotoapparat/selector/FocusModeSelectors.kt
package io.fotoapparat.selector

import io.fotoapparat.parameter.FocusMode

typealias FocusModeSelector = Iterable<FocusMode>.() -> FocusMode?

/**
 * @return Selector function which provides a non-adjustable focus mode if available.
 * Otherwise provides `null`.
 */
fun fixed(): FocusModeSelector = single(FocusMode.Fixed)

/**
 * @return Selector function which provides a focus mode targeting infinity if available.
 * Otherwise provides `null`.
 */
fun infinity(): FocusModeSelector = single(FocusMode.Infinity)

/**
 * @return Selector function which provides a macro focus mode if available.
 * Otherwise provides `null`.
 */
fun macro(): FocusModeSelector = single(FocusMode.Macro)

/**
 * @return Selector function which provides an auto focus mode if available.
 * Otherwise provides `null`.
 */
fun autoFocus(): FocusModeSelector = single(FocusMode.Auto)

/**
 * @return Selector function which provides a focus mode which constantly tries to stay
 * in focus if available. The speed of focus change is aggressive.
 * Otherwise provides `null`.
 */
fun continuousFocusPicture(): FocusModeSelector = single(FocusMode.ContinuousFocusPicture)

/**
 * @return Selector function which provides a focus mode which constantly tries to stay
 * in focus if available. The speed of focus change is smooth.
 * Otherwise provides `null`.
 */
fun continuousFocusVideo(): FocusModeSelector = single(FocusMode.ContinuousFocusVideo)

/**
 * @return Selector function which provides a focus mode which will produce images with
 * an extended depth of field if available.
 * Otherwise provides `null`.
 */
fun edof(): FocusModeSelector = single(FocusMode.Edof)

